# Exploratory Data Analysis (EDA) - Plotly Version

This notebook analyzes the education and literacy datasets prepared during the feature engineering step using **Plotly** for interactive data visualizations.
1. Average Years of Education
2. Illiteracy Rates
3. Literacy Rates

In [27]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
# import warnings

# warnings.filterwarnings('ignore')

## 1. Load the Datasets

In [28]:
# Define paths
path_avg = '../data/implemented_data/avg_years_of_education_p.csv'
path_ill = '../data/implemented_data/illeteracy_p.csv'
path_lit = '../data/implemented_data/literacy_p.csv'

# Load dataframes
df_avg = pd.read_csv(path_avg)
df_ill = pd.read_csv(path_ill)
df_lit = pd.read_csv(path_lit)

print("Data loaded successfully!")

Data loaded successfully!


## 2. Analysis: Average Years of Education (`df_edu`)

In [29]:
df_avg.head()

,COUNTRY,CODE,YEAR,GDP_PCAP,REGION,LITERACY_RATE,AVG_YEARS_OF_EDUCATION,POPULATION,GDP PER SCHOOLING YEAR,EDUCATION INDEX,EDUCATION LEVEL CATEGORY
0,AFGHANISTAN,AFG,1990,1617.8264,ASIA,82.254032,6.488154,8.388452e+07,249.350806,0.627542,MEDIUM
1,AFGHANISTAN,AFG,1995,1617.8264,ASIA,82.254032,6.488154,8.388452e+07,249.350806,0.627542,MEDIUM
2,AFGHANISTAN,AFG,2000,1617.8264,ASIA,82.254032,6.488154,8.388452e+07,249.350806,0.627542,MEDIUM
3,AFGHANISTAN,AFG,2001,1454.1108,ASIA,82.254032,6.488154,8.388452e+07,224.117804,0.627542,MEDIUM
4,AFGHANISTAN,AFG,2002,1774.3087,ASIA,82.254032,6.488154,8.388452e+07,273.468960,0.627542,MEDIUM


In [30]:
df_avg.describe()

,YEAR,GDP_PCAP,LITERACY_RATE,AVG_YEARS_OF_EDUCATION,POPULATION,GDP PER SCHOOLING YEAR,EDUCATION INDEX
count,6847.000000,6847.000000,6847.000000,6847.000000,6.847000e+03,6847.000000,6847.000000
mean,2007.181539,21885.927984,80.503999,6.491924,3.606163e+07,3091.558636,0.618917
std,10.053199,24067.136511,14.451406,1.873045,2.756797e+07,3209.684161,0.130986
min,1990.000000,510.822800,58.052015,4.201922,2.808810e+06,117.026016,0.430324
25%,1999.000000,4469.463750,58.052015,4.201922,1.810373e+07,819.416323,0.430324
50%,2007.000000,12636.182000,82.254032,6.488154,2.260825e+07,1975.495799,0.627542
75%,2016.000000,32038.773000,90.548342,6.662325,3.108848e+07,4331.410336,0.651806
max,2024.000000,174569.520000,97.149803,9.569358,8.388452e+07,26905.884614,0.804728


### Univariate Analysis (df_edu)

In [31]:
fig = make_subplots(rows=1, cols=3, subplot_titles=(
    'Literacy Distribution', 
    'AVG_YEARS_OF_EDUCATION Distribution', 
    'GDP_PER_CAPITA (Female) Distribution'))

fig.add_trace(go.Histogram(x=df_avg['LITERACY_RATE'], name='Literacy', marker_color='red', opacity=0.7), row=1, col=1)
fig.add_trace(go.Histogram(x=df_avg['AVG_YEARS_OF_EDUCATION'], name='Avg_years_of_education', marker_color='green', opacity=0.7), row=1, col=2)
fig.add_trace(go.Histogram(x=df_avg['GDP_PCAP'], name='GDP_per_capta', marker_color='blue', opacity=0.7), row=1, col=3)
fig.update_layout(title_text='Average years of education over GDP per capita Distributions', height=500, showlegend=True)
fig.show()

### Insights: Univariate Analysis (df_edu)
- **Average Years of Education**: Centered around 6.5 years globally, with a minimum of 4.2 and maximum of 9.5 years, showing significant disparity.
- **GDP Per Capita**: Highly skewed distribution. The median ($12,636) is much lower than the mean ($21,885) and the max ($174,569), highlighting global wealth inequalities.
- **Literacy Rate**: The median literacy rate is 82.2%, indicating that a majority of the global population has basic reading skills, though lower quartiles show room for improvement.

### Bivariate Analysis (df_edu)

In [32]:
fig = px.scatter(df_avg, x='GDP_PCAP', y='AVG_YEARS_OF_EDUCATION', color='REGION', 
                 log_x=True, hover_name='COUNTRY', hover_data=['YEAR'],
                 title='GDP per Capita (Log Scale) vs Average Years of Education',
                 opacity=0.7)
fig.show()

fig = px.scatter(df_avg, x='LITERACY_RATE', y='AVG_YEARS_OF_EDUCATION', color='REGION',
                 hover_name='COUNTRY', hover_data=['YEAR'],
                 title='Literacy Rate vs Average Years of Education',
                 opacity=0.7)
fig.show()

In [44]:
fig = px.scatter(df_avg, x='REGION', y='AVG_YEARS_OF_EDUCATION', color='REGION',
             title='Average Years of Education by Region')
fig.show()

In [34]:
# Correlation Matrix
numeric_cols = df_avg.select_dtypes(include=[np.number]).columns
corr_matrix = df_avg[numeric_cols].corr()

fig = px.imshow(corr_matrix, text_auto=True, aspect="auto", 
                title='Correlation Matrix for Education Dataset',
                color_continuous_scale='RdBu_r')
fig.show()

### Insights: Bivariate Analysis (df_edu)
- **Education & Literacy**: There is a strong positive correlation (r=0.89) between Average Years of Education and Literacy Rate, confirming that longer schooling directly translates to better literacy.
- **Wealth & Education**: GDP per Capita shows a notable positive correlation (r=0.49) with education years. Wealthier regions tend to provide longer education.
- **Education Index**: Shows a near-perfect correlation (r=0.96) with the average years of education, validating its use as a primary educational metric.

## 3. Analysis: Illiteracy (`df_ill`)

In [35]:
df_ill.head()

,COUNTRY,CODE,YEAR,ILLITERACY_RATE,LITERACY_RATE,ILLITERACY%,LITERACY GROWTH RATE,ILLITERACY TO LITERACY RATIO
0,AFGHANISTAN,AFG,2011,69.000000,31.00000,69.00,0.000000,2.225806
1,AFGHANISTAN,AFG,2015,66.246155,33.75384,66.25,8.883355,1.962626
2,AFGHANISTAN,AFG,2021,63.000000,37.00000,63.00,9.617158,1.702703
3,ALBANIA,ALB,2001,1.000000,99.00000,1.00,0.000000,0.010101
4,ALBANIA,ALB,2008,4.000000,96.00000,4.00,-3.030303,0.041667


In [36]:
df_ill.describe()

,YEAR,ILLITERACY_RATE,LITERACY_RATE,ILLITERACY%,LITERACY GROWTH RATE,ILLITERACY TO LITERACY RATIO
count,833.000000,833.000000,833.000000,833.000000,833.000000,833.000000
mean,2009.732293,17.943298,82.056702,17.943265,3.067580,0.369522
std,8.460407,19.816981,19.816981,19.816997,9.971740,0.733332
min,1990.000000,0.000000,11.000000,0.000000,-23.255814,0.000000
25%,2004.000000,4.000000,72.547130,4.000000,0.000000,0.041667
50%,2011.000000,9.000000,91.000000,9.000000,0.000000,0.098901
75%,2016.000000,27.452873,96.000000,27.450000,2.298851,0.378414
max,2023.000000,89.000000,100.000000,89.000000,136.363636,8.090909


### Univariate Analysis (df_ill)

In [37]:
fig1 = px.histogram(df_ill, x='ILLITERACY_RATE', nbins=30, marginal="violin",
                   title='Distribution of Illiteracy Rate', color_discrete_sequence=['purple'])
fig1.show()

fig2 = px.histogram(df_ill, x='ILLITERACY TO LITERACY RATIO', nbins=30, marginal="violin",
                   title='Distribution of Illiteracy:Literacy Ratio', color_discrete_sequence=['orange'])
fig2.show()

### Insights: Univariate Analysis (df_ill)
- **Illiteracy Rate**: Global average is ~17.9%, but the median is 9.0%. The right-skewed distribution indicates that while most countries have low illiteracy, a few still suffer from rates up to 89%.
- **Literacy Growth**: Positive growth (mean 3.06%) highlights successful global educational interventions over the years.
- **Illiteracy-to-Literacy Ratio**: Primarily concentrated near 0, but extends up to 8.09 in extreme cases.

### Bivariate Analysis (df_ill)

In [38]:
fig = px.scatter(df_ill, x='LITERACY_RATE', y='ILLITERACY_RATE', 
                 hover_name='COUNTRY', hover_data=['YEAR'],
                 title='Literacy Rate vs Illiteracy Rate', 
                 color_discrete_sequence=['darkgreen'], opacity=0.6)
fig.show()

### Insights: Bivariate Analysis (df_ill)
- **Illiteracy vs Literacy**: A perfect negative correlation (r=-1.0), as expected, since they are largely inversely proportional components.
- **Growth Rate Trends**: Literacy growth rate shows a slight positive correlation with illiteracy rate (r=0.30) - suggesting countries with higher illiteracy have more room for rapid growth.
- **Ratio Dependencies**: The illiteracy-to-literacy ratio strongly mirrors the illiteracy rate itself (r=0.85).

## 4. Analysis: Literacy gaps and demographics (`df_lit`)

In [39]:
df_lit.head()

,COUNTRY,CODE,YEAR,ADULT_LITERACY,YOUTH_LITERACY_M,YOUTH_LITERACY_F,REGION,LITERACY GAP GENDER,YOUTH LITERACY AVERAGE,ADULT VS YOUTH LITERACY GAP
0,AFGHANISTAN,AFG,2011,31.00000,62.00000,32.00000,ASIA,30.00000,47.000000,16.000000
1,AFGHANISTAN,AFG,2015,33.75384,57.73505,25.48416,ASIA,32.25089,41.609605,7.855765
2,AFGHANISTAN,AFG,2021,37.00000,71.00000,42.00000,ASIA,29.00000,56.500000,19.500000
3,AFGHANISTAN,AFG,2022,37.00000,83.40000,44.17171,ASIA,39.22829,63.785855,26.785855
4,ALBANIA,ALB,2001,99.00000,99.00000,99.00000,EUROPE,0.00000,99.000000,0.000000


In [40]:
df_lit.describe()

,YEAR,ADULT_LITERACY,YOUTH_LITERACY_M,YOUTH_LITERACY_F,LITERACY GAP GENDER,YOUTH LITERACY AVERAGE,ADULT VS YOUTH LITERACY GAP
count,1111.000000,1111.000000,1111.000000,1111.000000,1111.000000,1111.000000,1111.000000
mean,2010.090909,81.150331,91.292470,87.894665,4.842690,89.593567,8.443236
std,8.091249,20.031539,12.966018,19.408911,8.037869,15.914791,7.542108
min,1990.000000,11.000000,25.000000,9.000000,0.000000,17.500000,-9.793140
25%,2005.000000,71.000000,88.245000,84.700000,0.054730,86.000000,3.000000
50%,2011.000000,90.000000,97.600000,98.000000,1.000000,97.592790,6.500000
75%,2016.000000,96.000000,99.000000,99.000000,5.795735,99.000000,12.402500
max,2023.000000,100.000000,100.000000,100.000000,52.210000,100.000000,65.000000


### Univariate Analysis (df_lit)

In [41]:
fig = make_subplots(rows=1, cols=3, subplot_titles=(
    'Adult Literacy Distribution', 
    'Youth Literacy (Male) Distribution', 
    'Youth Literacy (Female) Distribution'))

fig.add_trace(go.Histogram(x=df_lit['ADULT_LITERACY'], name='Adult', marker_color='blue', opacity=0.7), row=1, col=1)
fig.add_trace(go.Histogram(x=df_lit['YOUTH_LITERACY_M'], name='Youth Male', marker_color='cyan', opacity=0.7), row=1, col=2)
fig.add_trace(go.Histogram(x=df_lit['YOUTH_LITERACY_F'], name='Youth Female', marker_color='magenta', opacity=0.7), row=1, col=3)

fig.update_layout(title_text='Literacy Distributions', height=500, showlegend=False)
fig.show()

### Insights: Univariate Analysis (df_lit)
- **Youth vs Adult Literacy**: Youth literacy (median 97.6% for males, 98.0% for females) is significantly higher than adult literacy (median 90.0%), proving generational progress in education.
- **Gender Gap**: The average gender literacy gap is 4.8%. While the median is a low 1.0%, extreme cases reach up to a 52.2% gap.
- **Adult vs Youth Gap**: The mean gap of 8.4% confirms recent historical acceleration in youth education access.

### Bivariate Analysis (df_lit)

In [42]:
fig = px.scatter(df_lit, x='YOUTH_LITERACY_M', y='YOUTH_LITERACY_F', color='REGION',
                 hover_name='COUNTRY', hover_data=['YEAR'],
                 title='Youth Literacy: Male vs Female', opacity=0.7)
# Add equality line
fig.add_shape(type='line', x0=0, y0=0, x1=100, y1=100, line=dict(color='red', dash='dash'))
fig.show()

fig = px.scatter(df_lit, x='YOUTH LITERACY AVERAGE', y='ADULT_LITERACY', color='REGION',
                 hover_name='COUNTRY', hover_data=['YEAR'],
                 title='Youth Literacy Avg vs Adult Literacy', opacity=0.7)
# Add equality line
fig.add_shape(type='line', x0=0, y0=0, x1=100, y1=100, line=dict(color='red', dash='dash'))
fig.show()

In [43]:
fig = px.box(df_lit, x='REGION', y='LITERACY GAP GENDER', color='REGION',
             title='Gender Literacy Gap by Region (Male - Female %)')
fig.show()

### Insights: Bivariate Analysis (df_lit)
- **Gender Parity via Overall Education**: There is a strong negative correlation (r=-0.79) between Adult Literacy and Gender Gap. As general literacy increases, gender disparities tend to shrink.
- **Youth Trajectory**: Male and female youth literacy are highly correlated (r=0.93). A country improving male youth education typically improves female youth education synchronously.
- **Intergenerational Growth**: Adult literacy strongly predicts youth literacy (r=0.93), showing cumulative educational foundations over generations.